### ISSUE TITLE
Turning on jbuilder for API only apps cause controller test failures

### ISSUE THREAD

**Issue by dhh** [rails/rails#25183]
Steps to replicate:

```
rails new robots --api
cd robots
rails generate scaffold post title:string
rails db:migrate
<Uncomment jbuilder gem from Gemfile>
bundle
rails test
```

Failures:

```
PostsControllerTest#test_should_update_post:
ActionView::MissingTemplate: Missing template posts/update, application/update with {:locale=>[:en], :formats=>[:html], :variants=>[], :handlers=>[:raw, :erb, :html, :builder, :ruby, :jbuilder]}. Searched in:
  * "/Users/david/Code/demos/robots/app/views"

    app/controllers/posts_controller.rb:30:in `update'
    test/controllers/posts_controller_test.rb:27:in `block in <class:PostsControllerTest>'
```

Line 30 has `render json: @post`, so its not relying on default rendering. It should just be calling `#to_json` on `@post` and return that. Shouldn't matter what rendering engines are installed or not.

**Comment by dhh** [comment]
cc @rafaelfranca

**Comment by maclover7** [comment]
I think we just need to bump the version of `jbuilder` here, there [was an issue](https://github.com/rails/jbuilder/issues/317) with `jbuilder` and `ActionController::API`, but it looks like it's fixed [in this commit](https://github.com/rails/jbuilder/commit/64c510ec69d9e63b73ffd5942e802d21a7d14701).

**Comment by prathamesh-sonpatki** [comment]
Confirm that jbuilder master fixes this issue. We need a release for jbuilder.

**Comment by dhh** [comment]
@rwz Could you cut a new release to resolve this issue, Pavel? Thanks!

**Comment by rwz** [comment]
@dhh done. You might wanna depend on jbuilder ~> 2.5 by default in rails 5.

**Comment by vipulnsward** [comment]
Closing https://rubygems.org/gems/jbuilder/versions/2.5.0

**Comment by dhh** [comment]
Thanks @rwz! I bumped the dependency to 2.5 in 998bbbc.

**Comment by y-yagi** [comment]
Is this really fixed?
I tried in Rails 5 and jbuilder 2.5.0, but it seems to still be in error.

Steps to reproduce:

```
rails new robots --api
cd robots
<Uncomment jbuilder gem from Gemfile>
bundle
bin/rails generate scaffold post title:string
bin/rails db:migrate
bin/rails test
```

Failures:

```
E

Error:
PostsControllerTest#test_should_get_index:
ActionController::UnknownFormat: PostsController#index is missing a template for this request format and variant.

request.formats: ["text/html"]
request.variant: []
    test/controllers/posts_controller_test.rb:9:in `block in <class:PostsControllerTest>'


bin/rails test test/controllers/posts_controller_test.rb:8 
```
I think need to specify the format in test code?

**Comment by dhh** [comment]
Replicate @y-yagi's issue. This should work.

**Comment by dhh** [comment]
@rwz @rafaelfranca Thoughts on this?

**Comment by rafaelfranca** [comment]
I'll take a look. It is really a weird behavior.

**Comment by prathamesh-sonpatki** [comment]
This is failing because by default the request encoder  is HTML - https://github.com/prathamesh-sonpatki/rails/blob/master/actionpack/lib/action_dispatch/testing/integration.rb#L327

We can fix it in following way by generating API controller tests with `as: :json` as follows.

``` ruby
test "should get index" d 
    get posts_url, as: :json
    assert_response :success
  end
```

cc @kaspth

**Comment by dhh** [comment]
Ah yes of course. Seems like jbuilder needs to provide controller test stubs on its own.

**Comment by prathamesh-sonpatki** [comment]
@dhh Tried to fix it here https://github.com/rails/rails/issues/25317.

**Comment by sgrif** [comment]
See https://github.com/rails/rails/pull/25317#issuecomment-224376800, this issue isn't fully resolved. There's a bandaid in place for 5.0.0 though, so I'm moving this to the 5.0.1 milestone.

**Comment by baurine** [comment]
I met the same problem with Rails 5.0.1 + RSpec 3.5.2 + JBuilder 2.6.0.

the failure log:

```
Failures:

  1) API::SessionsController POST #create should return succeed json with valid params
     Failure/Error: post :create, params: correct_params

     ActionController::UnknownFormat:
       API::SessionsController#create is missing a template for this request format and variant.

       request.formats: ["text/html"]
       request.variant: []
```

I work around this problem by manually specify the request header accept. reference from a [stackoverflow answer](http://stackoverflow.com/a/27968789/2998877).

```ruby
  before :each do
    request.headers['accept'] = 'application/json'
  end
```

**Comment by maclover7** [comment]
Opened PR to try and solve this --> #27333

**Comment by aried3r** [comment]
I tried this with Ruby 2.6.3 and Rails 6.0.0.rc1, following the steps in the beginning of this issue. It seems to be fixed at least in this version, but the jbuilder version defined in `Gemfile` allows jbuilder versions that won't even install, I'm not sure if it's too late to change this in the `Gemfile` template? I mean, bundler will make sure that it won't be used, but `~> 2.5` is also not correct.

jbuilder 2.9.1
```console
$ rails test
Run options: --seed 10238

# Running:

.....

Finished in 0.310141s, 16.1217 runs/s, 22.5704 assertions/s.
5 runs, 7 assertions, 0 failures, 0 errors, 0 skips
```

jbuilder 2.8.0
```console
$ rails test
DEPRECATION WARNING: Single arity template handlers are deprecated. Template handlers must
now accept two parameters, the view object and the source for the view object.
Change:
  >> JbuilderHandler.call(template)
To:
  >> JbuilderHandler.call(template, source)
 (called from <main> at /private/tmp/robots/config/environment.rb:5)
Run options: --seed 12734

# Running:

.....

Finished in 0.288970s, 17.3028 runs/s, 24.2240 assertions/s.
5 runs, 7 assertions, 0 failures, 0 errors, 0 skips
```

jbuilder 2.7
```console
$ rails test
DEPRECATION WARNING: Single arity template handlers are deprecated. Template handlers must
now accept two parameters, the view object and the source for the view object.
Change:
  >> JbuilderHandler.call(template)
To:
  >> JbuilderHandler.call(template, source)
 (called from <main> at /private/tmp/robots/config/environment.rb:5)
Run options: --seed 45429

# Running:

.....

Finished in 0.244076s, 20.4854 runs/s, 28.6796 assertions/s.
5 runs, 7 assertions, 0 failures, 0 errors, 0 skips
```

jbuilder 2.6
```console
$ bundle install
Fetching gem metadata from https://rubygems.org/............
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Bundler could not find compatible versions for gem "activesupport":
  In snapshot (Gemfile.lock):
    activesupport (= 6.0.0.rc1)

  In Gemfile:
    jbuilder (= 2.6) was resolved to 2.6.0, which depends on
      activesupport (>= 3.0.0, < 5.1)

    rails (~> 6.0.0.rc1) was resolved to 6.0.0.rc1, which depends on
      activesupport (= 6.0.0.rc1)

    spring was resolved to 2.0.2, which depends on
      activesupport (>= 4.2)

    rails (~> 6.0.0.rc1) was resolved to 6.0.0.rc1, which depends on
      sprockets-rails (>= 2.0.0) was resolved to 3.2.1, which depends on
        activesupport (>= 4.0)

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.
```

jbuilder 2.5 which is the lowest version Rails supports (`~> 2.5`), manually setting it to `2.5` and running `bundle install`:
```console
$ bundle install
Fetching gem metadata from https://rubygems.org/............
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Bundler could not find compatible versions for gem "activesupport":
  In snapshot (Gemfile.lock):
    activesupport (= 6.0.0.rc1)

  In Gemfile:
    jbuilder (= 2.5) was resolved to 2.5.0, which depends on
      activesupport (>= 3.0.0, < 5.1)

    rails (~> 6.0.0.rc1) was resolved to 6.0.0.rc1, which depends on
      activesupport (= 6.0.0.rc1)

    spring was resolved to 2.0.2, which depends on
      activesupport (>= 4.2)

    rails (~> 6.0.0.rc1) was resolved to 6.0.0.rc1, which depends on
      sprockets-rails (>= 2.0.0) was resolved to 3.2.1, which depends on
        activesupport (>= 4.0)

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.
```

**Comment by aried3r** [comment]
I believe this can be closed.
